### ISSUE TITLE
Quantile Regression

### ISSUE THREAD

**Issue by david-waterworth** [catboost/catboost#719]
I'd like to investigate quantile regression, i.e. Qalpha[y|X] instead of E[y|X] 

sklearn's `GradientBoostingRegressor` has `quantile` as a loss function.  For each loss function there's a class which implements `__call__()`, `negative_gradient()` and `update_terminal_regions()` which implements the loss function, it's first derivative and the associated leaf predictor repectively (i.e. L2 loss has uses `mean()` for the predictor, L1 `median()` and quantile uses the `percentile(alpha)` across the leaf targets.

I've looked at the custom loss examples in catboost, from what I see they require first and second derivative functions but I don't see a way to replace the leaf prediction, is this possible (ideally initially as python, then if the result in promising I'll attempt a C++ implementation)

Also in general, I don't see anything in the documentation indicating which leaf prediction function is used for different loss functions? Are you using mean for RMSE and median for MAE?

Also I just found this https://github.com/catboost/catboost/issues/37 which implies you've implementation quantile loss even if it's not exposed via the api? Is that the case? Is there an easy way I can change alpha from 0.5 to 0.95 say?

**Comment by ian-contiamo** [comment]
I can only answer your last question. You can specify the quantile loss as follows:

`model = catboost.CatBoostRegressor(loss_function='Quantile:alpha=0.95', ...)`

**Comment by david-waterworth** [comment]
Thanks @ian-contiamo, when I fitted `GradientBoostingRegressor(loss='quantile', alpha=0.95,...)` to the residuals of my base CatBoost mode I got a prediction which looks quite reasonable as an estimate of 95th percentile. When I used `CatBoostRegressor(loss_function='Quantile:alpha=0.95', ...)` the prediction appears to be the mean as it was close to zero so it doesn't appear to be adjusting the predictor function to be consistent with the loss which is odd. Hopefully @annaveronika  can shed some light?

**Comment by david-waterworth** [comment]
OK I think I've got to the bottom of this - quantile regression does work, but it converges very slowly if at all.

It's likely related to https://github.com/Microsoft/LightGBM/pull/1199, there's a good description [here](http://jmarkhou.com/lgbqr/). 

I'm not 100% sure, but if the leaf values are approximated by L'(X,y) / L''(X,y) then it's no surprise that it doesn't work so well for the quantile loss with high / low alpha due to the zero 2nd derivative.

Specifically I fitted the model

```
X = np.array([1,1,1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2]).reshape(-1, 1)
y = np.array([1,2,3,4,5,6,7,8,9,10,21,22,23,24,25,26,27,28,29,30])
```

i.e. there's two clusters labeled 1 and 2, with target mean 5.5 and 28.5, and I attempted to fit using Quantile:alpha=0.95, I set learning_rate, max_depth to 1 and experimented with iterations. 

Both lightgbm and sklearn's GradientBoostingRegressor converge in a single iteration to a solution y=8|X=1 y=28|X=2 (GradientBoostingRegressor) or y=8.5|X=1 y=28.5|X=2 (lightgbm)

catboost takes around 100 iterations to converge. The table below shows that after the first iteration all the leaf estimates are the same, and there's a large error (it appears that there's no split, I think it has split at the correct place though - I tried to verify using the standalone_evaluator but I couldn't easily debug the stl containers using vscode). At each step the estimates improve very slowly. I tried to set --leaf-estimation-method Newton but this isn't supported. Are there other parameters I can try?

![image](https://user-images.githubusercontent.com/5028974/54170390-98412700-44ca-11e9-99ce-ea9497ced68e.png)

**Comment by david-waterworth** [comment]
I'm (slowly) starting to understand what's happening here. In this case, since the loss is quantile, Newton's method isn't used to update the leafs so the comment above doesn't really apply, Gradient method is used.

What's actually happening is the initial approx is zero. This means the derivs for every document at iteration 1 are equal to alpha (0.75) due to symmetry of the quantile loss function. The delta approximation is then alpha / (10+3) where 10 is the number of values per leaf (since it split down the middle) and 3 is the L2 normalisation. So after a single iteration both leaves have value of 0.58

After a few iterations, we start to get -ve errors for some of the leaves and then the derivatives start to diverge. 

The issue seems to be the poor initial approximation. I might experiment with starting from the mean to see if that provides more direction.

**Comment by annaveronika** [comment]
We have implemented Exact method for Quantile regression and also starting from best constant approximation for this mode.
Exact method is available on CPU only for now, and is set as a default method.
Starting from best approximation is off by default for now, but will be on in one of the next releases. For now please use boost_from_average=True flag for it (this name is used for compatability with other GBDT libraries).

And thank you very much for your input, we really appreciate it!

**Comment by annaveronika** [comment]
Closing this issue, and keep updated, we will post in in release notices, when boost_from_average is True by default.

**Comment by david-waterworth** [comment]
Thanks @annaveronika Note if you set `leaf_estimation_method=Exact` and `task_type=GPU` there's no warning, it (I guess) quietly falls back to Newton.

**Comment by annaveronika** [comment]
> Thanks @annaveronika Note if you set `leaf_estimation_method=Exact` and `task_type=GPU` there's no warning, it (I guess) quietly falls back to Newton.

Thanks for the report! We'll add an exception for this case.

**Comment by Sandy4321** [comment]
any updates, is catboos better now for quantile regression than lightgbm ?
https://github.com/microsoft/LightGBM/pull/1199

https://github.com/microsoft/LightGBM/issues/1182
I think now with the new updates that fix this issue, LightGBM is the fastest, quantiles-supporting boosted decision tree implementation available. Pretty exciting! I'm getting about 20x speedup with similar performance over sklearn in quantile workloads! Great work.
